DATE_TRUNC result is incorrectly localized when a timezone is set

When no timezone is active, hour is zero as expected:
```sql
> CREATE TABLE "timezones_event" ("id" integer NOT NULL PRIMARY KEY DEFAULT unique_rowid(), "dt" timestamptz NOT NULL);
> INSERT INTO "timezones_event" ("dt") VALUES ('2011-01-01T01:30:00+00:00'::timestamptz);
> SELECT DISTINCT DATE_TRUNC('year', "timezones_event"."dt") AS "datetimefield" FROM "timezones_event" WHERE "timezones_event"."dt" IS NOT NULL ORDER BY "datetimefield" ASC;
        datetimefield        
+---------------------------+
  2011-01-01 00:00:00+00:00 
```
But when a timezone is active ('Africa/Nairobi' is UTC+3), the hour of the result is localized:
```sql
> SET TIME ZONE 'Africa/Nairobi';
> SELECT DISTINCT DATE_TRUNC('year', "timezones_event"."dt") AS "datetimefield" FROM "timezones_event" WHERE "timezones_event"."dt" IS NOT NULL ORDER BY "datetimefield" ASC;
        datetimefield        
+---------------------------+
  2011-01-01 03:00:00+03:00
```
I believe this is a bug in cockroachdb.

The relevant Django test:
```
======================================================================
FAIL: test_query_datetimes (timezones.tests.LegacyDatabaseTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/timezones/tests.py", line 180, in test_query_datetimes
    self.assertSequenceEqual(Event.objects.datetimes('dt', 'year'), [datetime.datetime(2011, 1, 1, 0, 0, 0)])
AssertionError: Sequences differ: <QuerySet [datetime.datetime(2010, 1, 1, 3, 0)]> != [datetime.datetime(2011, 1, 1, 0, 0)]

First differing element 0:
datetime.datetime(2010, 1, 1, 3, 0)
datetime.datetime(2011, 1, 1, 0, 0)

- <QuerySet [datetime.datetime(2010, 1, 1, 3, 0)]>
+ [datetime.datetime(2011, 1, 1, 0, 0)]
```

--- a/pkg/sql/logictest/testdata/logic_test/datetime
+++ b/pkg/sql/logictest/testdata/logic_test/datetime
@@ -721,27 +721,27 @@ FROM ex WHERE date_trunc_result IS NOT NULL ORDER BY k
 query T
 SELECT date_trunc('millennia', '2000-02-10 19:46:33.306157519-04'::timestamptz)::string
 ----
-1001-01-01 00:00:00-04:00
+1001-01-01 00:00:00+00:00
 
 query T
 SELECT date_trunc('centuries', '2016-02-10 19:46:33.306157519-04'::timestamptz)::string
 ----
-2001-01-01 00:00:00-04:00
+2001-01-01 00:00:00+00:00
 
 query T
 SELECT date_trunc('decades', '2016-02-10 19:46:33.306157519-04'::timestamptz)::string
 ----
-2010-01-01 00:00:00-04:00
+2010-01-01 00:00:00+00:00
 
 query T
 SELECT date_trunc('hour', '2016-02-10 19:46:33.306157519-04'::timestamptz)::string
 ----
-2016-02-10 19:00:00-04:00
+2016-02-10 23:00:00+00:00
 
 query T
 SELECT date_trunc('hours', '2016-02-10 19:46:33.306157519-04'::timestamptz)::string
 ----
-2016-02-10 19:00:00-04:00
+2016-02-10 23:00:00+00:00
 
 query IBT
 SELECT k, date_trunc(element, input::date) = date_trunc_result::date, date_trunc(element, input::date)::string
@@ -1363,6 +1363,31 @@ SELECT date_trunc('month', "date") AS date_trunc_month_created_at FROM "topics";
 ----
 2017-12-01 00:00:00 +0000 UTC
 
+# Test date_trunc works when timestamp zone changes.
+subtest regression_41663
+
+query T
+select date_trunc('day', '2011-01-01 22:30:00'::date);
+----
+2011-01-01 00:00:00 +0000 UTC
+
+query T
+select date_trunc('day', '2011-01-01 22:30:00+01:00'::timestamptz);
+----
+2011-01-01 00:00:00 +0000 UTC
+
+statement ok
+SET TIME ZONE 'Africa/Nairobi';
+
+query T
+select date_trunc('day', '2011-01-01 22:30:00'::date);
+----
+2011-01-01 00:00:00 +0300 +0300
+
+query T
+select date_trunc('day', '2011-01-01 22:30:00+01:00'::timestamptz);
+----
+2011-01-02 00:00:00 +0300 +0300
 
 # Test negative years to ensure they can round-trip through the parser.
 # Also ensure that we don't trigger any of the "convenience" rules.

--- a/pkg/sql/sem/builtins/builtins.go
+++ b/pkg/sql/sem/builtins/builtins.go
@@ -1954,7 +1954,7 @@ may increase either contention or retry errors, or both.`,
 			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
 				fromTSTZ := args[1].(*tree.DTimestampTZ)
 				timeSpan := strings.ToLower(string(tree.MustBeDString(args[0])))
-				return truncateTimestamp(ctx, fromTSTZ.Time, timeSpan)
+				return truncateTimestamp(ctx, fromTSTZ.Time.In(ctx.GetLocation()), timeSpan)
 			},
 			Info: "Truncates `input` to precision `element`.  Sets all fields that are less\n" +
 				"significant than `element` to zero (or one, for day and month)\n\n" +

--- a/pkg/sql/sem/builtins/builtins_test.go
+++ b/pkg/sql/sem/builtins/builtins_test.go
@@ -15,11 +15,15 @@ import (
 	"fmt"
 	"math/rand"
 	"testing"
+	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestCategory(t *testing.T) {
@@ -239,6 +243,40 @@ func TestLPadRPad(t *testing.T) {
 		if out != tc.expected {
 			t.Errorf("expected %s, found %s", tc.expected, out)
 		}
+	}
+}
+
+func TestTruncateTimestamp(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	loc, err := timeutil.LoadLocation("Australia/Sydney")
+	require.NoError(t, err)
+
+	testCases := []struct {
+		fromTime time.Time
+		timeSpan string
+		expected *tree.DTimestampTZ
+	}{
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "millennium", tree.MakeDTimestampTZ(time.Date(2001, time.January, 1, 0, 0, 0, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "century", tree.MakeDTimestampTZ(time.Date(2101, time.January, 1, 0, 0, 0, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "decade", tree.MakeDTimestampTZ(time.Date(2110, time.January, 1, 0, 0, 0, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "year", tree.MakeDTimestampTZ(time.Date(2118, time.January, 1, 0, 0, 0, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "quarter", tree.MakeDTimestampTZ(time.Date(2118, time.January, 1, 0, 0, 0, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "month", tree.MakeDTimestampTZ(time.Date(2118, time.March, 1, 0, 0, 0, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "day", tree.MakeDTimestampTZ(time.Date(2118, time.March, 11, 0, 0, 0, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "week", tree.MakeDTimestampTZ(time.Date(2118, time.March, 6, 0, 0, 0, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "hour", tree.MakeDTimestampTZ(time.Date(2118, time.March, 11, 5, 0, 0, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "second", tree.MakeDTimestampTZ(time.Date(2118, time.March, 11, 5, 6, 7, 0, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "millisecond", tree.MakeDTimestampTZ(time.Date(2118, time.March, 11, 5, 6, 7, 80000000, loc), time.Microsecond)},
+		{time.Date(2118, time.March, 11, 5, 6, 7, 80009001, loc), "microsecond", tree.MakeDTimestampTZ(time.Date(2118, time.March, 11, 5, 6, 7, 80009000, loc), time.Microsecond)},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.timeSpan, func(t *testing.T) {
+			result, err := truncateTimestamp(nil, tc.fromTime, tc.timeSpan)
+			require.NoError(t, err)
+			assert.Equal(t, tc.expected, result)
+		})
 	}
 }
 

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -2124,7 +2124,7 @@ func MakeDTimestampTZFromDate(loc *time.Location, d *DDate) (*DTimestampTZ, erro
 	if err != nil {
 		return nil, err
 	}
-	return MakeDTimestampTZ(t, time.Microsecond), nil
+	return MakeDTimestampTZ(t.In(loc), time.Microsecond), nil
 }
 
 // ParseDTimestampTZ parses and returns the *DTimestampTZ Datum value represented by

